src/imitation/algorithms/preference_comparisons.py

```diff
@@ -1138,30 +1224,101 @@ def _make_data_loader(self, dataset: PreferenceDataset) -> data_th.DataLoader:
             collate_fn=preference_collate_fn,
         )
 
+    @property
+    def requires_regularizer_update(self) -> bool:
+        """Whether the regularizer requires updating.
+
+        Returns:
+            If true, this means that a validation dataset will be used.
+        """
+        return self.regularizer is not None and self.val_split > 0
+
     def _train(self, dataset: PreferenceDataset, epoch_multiplier: float = 1.0) -> None:
         """Trains for `epoch_multiplier * self.epochs` epochs over `dataset`."""
-        dataloader = self._make_data_loader(dataset)
+        if self.requires_regularizer_update:
+            val_length = int(len(dataset) * self.val_split)
+            train_length = len(dataset) - val_length
+            if val_length < 1 or train_length < 1:
+                raise ValueError(
+                    "Not enough data samples to split into training and validation, "
+                    "or the validation split is too large/small. "
+                    "Make sure you've generated enough initial preference data. "
+                    "You can adjust this through initial_comparison_frac in "
+                    "PreferenceComparisons.",
+                )
+            train_dataset, val_dataset = data_th.random_split(
+                dataset,
+                lengths=[train_length, val_length],
+                generator=th.Generator().manual_seed(self.seed),
+            )
+            dataloader = self._make_data_loader(train_dataset)
+            val_dataloader = self._make_data_loader(val_dataset)
+        else:
+            dataloader = self._make_data_loader(dataset)
+            val_dataloader = None
+
         epochs = round(self.epochs * epoch_multiplier)
 
-        for _ in tqdm(range(epochs), desc="Training reward model"):
-            for fragment_pairs, preferences in dataloader:
-                self.optim.zero_grad()
-                loss = self._training_inner_loop(fragment_pairs, preferences)
-                loss.backward()
-                self.optim.step()
+        assert epochs > 0, "Must train for at least one epoch."
+        epoch_num = 0
+        with self.logger.accumulate_means("reward"):
+            for epoch_num in tqdm(range(epochs), desc="Training reward model"):
+                prefix = f"epoch-{epoch_num}"
+                train_loss = 0.0
+                for fragment_pairs, preferences in dataloader:
+                    self.optim.zero_grad()
+                    loss = self._training_inner_loop(
+                        fragment_pairs,
+                        preferences,
+                        mode=f"{prefix}/train",
+                    )
+                    train_loss += loss.item()
+                    if self.regularizer:
```

Comment:
We're not doing `loss.backward()` here. So if `self.regularizer` is `None`, we never make a backward call?

Unless I'm missing something, this seems like a major bug. We should probably add a test to catch this...

This also makes me more unhappy about `regularize` being responsible for calling `backward()` though I don't see a good alternative admittedly.

Reply:
The problem is that you apparently have to call backward before regularizing if you use weight decay so that the change in weights does not affect the gradient. Potential solutions:

1. Keep as is, add an `else: loss.backward()`
2. Do 1 but rename `.regularize()` to `.regularize_and_backward()` to make sure we don't miss this
3. Split `.regularize()` into `.pre_regularize()` and `.post_regularize()`, calling manually the `.backward()` method in the middle.

We should add the test to catch this anyways, but super confused as to why this is not failing. If there is no gradient in the tensors, how is the optimizer doing gradient descent?

Reply:
OTOH 2 seems like the best solution. I think 3 could be better if there's something we'd want to do (other than backward the loss) in between "pre" and "post" regularization -- but otherwise it's just adding extra cruft, and we can still forget to backward.

The grad parameter is still present without backward, just zero (set by `zero_grad()`). So won't be caught by tests unless we explicitly check for loss going down.

Reply:
Is there a way to add to all our tests where we use an optimizer that the loss is going down? This feels like a pretty big bug that we should systematically test for.

Reply:
But the loss doesn't always go down -- at least not in SGD, for small batch sizes.

You can absolutely test this but you have to be careful to train for long enough, set sufficiently conservative thresholds, etc. Don't think it's something we can just tack on as boilerplate.